lua/nvim-tree/git/cli.lua

```diff
@@ -0,0 +1,136 @@
+local notify = require "nvim-tree.notify"
+local log = require "nvim-tree.log"
+-- vim.loop got renamed in 0.10.0
+-- ref: https://github.com/neovim/neovim/pull/22846
+local uv = vim.loop or vim.uv
+
+local M = {}
+
+-- Loggers
+M.notify = notify
+M.log = log
+
+--- See h `uv.spawn-options` for full list of props
+--- @class vim.uv.SpawnOptions
+--- @field args string[] args to be passed to git command e.g. -C
+--- @field env table|nil
+--- @field cwd table|nil
+--- @field detached boolean|nil detach process from parent
+--- @field timeout number|nil timeout before hanged git process gets killed
+
+--- A very lightweight generic git cli command wrapper powered by libuv;
+--- it's probably worth to consider to import another library
+--- @param uv_spawn_opts vim.uv.SpawnOptions uv.spaw like options
+--- @param on_err_or_data fun(err: string|nil, data: string|nil) error handler callback
+--- @param on_exit? fun(code: number, signal: number) error handler callback
+--- @param on_timeout? fun(): nil called if process timedout
+function M.cli(uv_spawn_opts, on_err_or_data, on_exit, on_timeout)
+  if (not M._GIT_IN_PATH) and vim.fn.executable "git" == 0 then
+    M.notify.error "nvim-tree: git not in path. Install git first. Aborting."
+    return
+  end
+  M._GIT_IN_PATH = true
+
+  if type(uv_spawn_opts) ~= "table" then
+    error("uv_spawn_opts table is required", 2)
+  end
+
+  if type(uv_spawn_opts.args) ~= "table" then
+    error("uv_spawn_opts.args table is required", 2)
+  end
+
+  if type(uv_spawn_opts.timeout) ~= "number" then
+    uv_spawn_opts.timeout = 10000 -- 8 seconds
```

Comment:
Please don't hardcode this, use `git.timeout`.

Runner has it and should pass it through:
```
[2023-11-26 14:17:01] [git] git -C /home/alex/src/linux rev-parse --show-toplevel --absolute-git-dir
/home/alex/src/linux
/home/alex/src/linux/.git
[2023-11-26 14:17:01] [git] git -C /home/alex/src/linux config status.showUntrackedFiles
[2023-11-26 14:17:01] [git] running job with timeout 400ms
[2023-11-26 14:17:01] [git] git --no-optional-locks status --porcelain=v1 -z --ignored=matching -u 400
[2023-11-26 14:17:01] [git] running job with timeout 10000ms
[2023-11-26 14:17:01] [git] git --no-optional-locks status --porcelain=v1 -z --ignored=matching -u
 M .clang-format
 M .cocciconfig
 D .get_maintainer.ignore
 D .gitattributes
 M .mailmap
 M .rustfmt.toml
 M COPYING
 D CREDITS
 M Kconfig
```

Reply:
[An ordinary argument type check](https://github.com/nvim-tree/nvim-tree.lua/pull/2551#discussion_r1405387603). Remove?

Reply:
Options are defaulted in the Runner's constructor.

Reply:
What do you mean? This is a separate module. It's surely going to be called elsewhere.